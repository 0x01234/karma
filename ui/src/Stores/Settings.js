import { localStored } from "mobx-stored";

import { action } from "mobx";

const defaultSavedFilters = {
  filters: [],
  present: false
};

class Settings {
  savedFilters = localStored("savedFilters", defaultSavedFilters, {
    delay: 100
  });

  saveFilters = action(newFilters => {
    this.savedFilters.filters = newFilters;
    this.savedFilters.present = true;
  });

  clearSavedFilters = action(() => {
    this.savedFilters.filters = [];
    this.savedFilters.present = false;
  });
}

export { Settings };
