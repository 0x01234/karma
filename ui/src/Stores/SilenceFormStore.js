import { observable, action, computed } from "mobx";

import uniqueId from "lodash.uniqueid";

import moment from "moment";

const NewEmptyMatcher = id => {
  return {
    id: uniqueId(),
    name: "",
    values: [],
    suggestions: {
      names: [],
      values: []
    },
    isRegex: false
  };
};

const ValueToObject = value => ({ label: value, value: value });

class SilenceFormStore {
  // this is used to store modal visibility toggle
  toggle = observable(
    {
      visible: false,
      toggle() {
        this.visible = !this.visible;
      },
      hide() {
        this.visible = false;
      },
      show() {
        this.visible = true;
      }
    },
    { toggle: action.bound, hide: action.bound, show: action.bound }
  );

  // form data is stored here, it's global (rather than attached to the form)
  // so it can be manipulated from other parts of the code
  // example: when user clicks a silence button on alert we should populate
  // this form from that alert so user can easily silence that alert
  data = observable(
    {
      inProgress: false,
      alertmanagers: [],
      matchers: [],
      startsAt: moment(),
      endsAt: moment().add(1, "hour"),
      comment: "",
      author: "",

      resetProgress() {
        this.inProgress = false;
      },

      // append a new empty matcher to the list
      addEmptyMatcher() {
        let m = NewEmptyMatcher();
        this.matchers.push(m);
      },

      deleteMatcher(id) {
        // only delete matchers if we have more than 1
        if (this.matchers.length > 1) {
          this.matchers = this.matchers.filter(m => m.id !== id);
        }
      },

      fillMatchersFromGroup(group) {
        let matchers = [];

        // add matchers for all shared labels in this group
        for (const [key, value] of Object.entries(
          Object.assign({}, group.labels, group.shared.labels)
        )) {
          matchers.push({
            id: uniqueId(),
            name: key,
            values: [ValueToObject(value)],
            suggestions: {
              names: [],
              values: []
            },
            isRegex: false
          });
        }

        // add matchers for all unique labels in this group
        let labels = {};
        for (const alert of group.alerts) {
          for (const [key, value] of Object.entries(alert.labels)) {
            if (!labels[key]) {
              labels[key] = new Set();
            }
            labels[key].add(value);
          }
        }
        for (const [key, values] of Object.entries(labels)) {
          matchers.push({
            id: uniqueId(),
            name: key,
            values: [...values].sort().map(value => ValueToObject(value)),
            suggestions: {
              names: [],
              values: []
            },
            isRegex: values.size > 1
          });
        }

        this.matchers = matchers;
      },

      verifyStarEnd() {
        if (this.endsAt.isSameOrBefore(this.startsAt)) {
          this.endsAt = moment(this.startsAt).add(1, "minutes");
        }
      },
      incStart(minutes) {
        this.startsAt = moment(this.startsAt).add(minutes, "minutes");
        this.verifyStarEnd();
      },
      decStart(minutes) {
        this.startsAt = moment(this.startsAt).subtract(minutes, "minutes");
        this.verifyStarEnd();
      },

      incEnd(minutes) {
        this.endsAt = moment(this.endsAt).add(minutes, "minutes");
        this.verifyStarEnd();
      },
      decEnd(minutes) {
        this.endsAt = moment(this.endsAt).subtract(minutes, "minutes");
        this.verifyStarEnd();
      },

      incDuration(minutes) {
        this.endsAt = moment(this.endsAt).add(minutes, "minutes");
      },
      decDuration(minutes) {
        const newEndsAt = moment(this.endsAt).subtract(minutes, "minutes");
        if (newEndsAt.isAfter(this.startsAt)) {
          this.endsAt = newEndsAt;
        }
      },

      get toAlertmanagerPayload() {
        const payload = {
          matchers: this.matchers.map(m => ({
            name: m.name,
            value:
              m.values.length > 1
                ? `(${m.values.map(v => v.value).join("|")})`
                : m.values.length === 1
                  ? m.values[0].value
                  : "",
            isRegex: m.isRegex
          })),
          startsAt: this.startsAt
            .second(0)
            .millisecond(0)
            .toISOString(),
          endsAt: this.endsAt
            .second(59)
            .millisecond(0)
            .toISOString(),
          createdBy: this.author,
          comment: this.comment
        };
        return payload;
      },
      get toDuration() {
        const data = {
          days: this.endsAt.diff(this.startsAt, "days"),
          hours: this.endsAt.diff(this.startsAt, "hours") % 24,
          minutes: this.endsAt.diff(this.startsAt, "minutes") % 60
        };
        return data;
      }
    },
    {
      resetProgress: action.bound,
      addEmptyMatcher: action.bound,
      deleteMatcher: action.bound,
      fillMatchersFromGroup: action.bound,
      verifyStarEnd: action.bound,
      incStart: action.bound,
      decStart: action.bound,
      incEnd: action.bound,
      decEnd: action.bound,
      incDuration: action.bound,
      decDuration: action.bound,
      toAlertmanagerPayload: computed,
      toDuration: computed
    },
    { name: "Silence form store" }
  );
}

export { SilenceFormStore };
