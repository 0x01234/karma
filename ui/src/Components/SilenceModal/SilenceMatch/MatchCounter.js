import React, { Component } from "react";
import PropTypes from "prop-types";

import { observable, action } from "mobx";
import { observer } from "mobx-react";

import { throttle } from "lodash";

import hash from "object-hash";

import { FontAwesomeIcon } from "@fortawesome/react-fontawesome";
import { faExclamationCircle } from "@fortawesome/free-solid-svg-icons/faExclamationCircle";

import { FormatQuery, QueryOperators, StaticLabels } from "Common/Query";
import { FormatBackendURI, FormatAlertsQ } from "Stores/AlertStore";
import { SilenceFormStore } from "Stores/SilenceFormStore";
import { SilenceFormMatcher } from "Models/SilenceForm";

const MatchCounter = observer(
  class MatchCounter extends Component {
    static propTypes = {
      silenceFormStore: PropTypes.instanceOf(SilenceFormStore).isRequired,
      matcher: SilenceFormMatcher.isRequired
    };

    matchedAlerts = observable(
      {
        total: 0,
        error: null,
        fetch: null,
        setTotal(value) {
          this.total = value;
        },
        setError(value) {
          this.error = value;
        }
      },
      {
        setTotal: action,
        setError: action
      }
    );

    onFetch = throttle(() => {
      const { silenceFormStore, matcher } = this.props;

      const filters = [];

      // append current matcher values as a filter
      const operator = matcher.isRegex
        ? QueryOperators.Regex
        : QueryOperators.Equal;
      const value =
        matcher.values.length > 1
          ? `(${matcher.values.map(v => v.value).join("|")})`
          : matcher.values[0].value;
      filters.push(
        FormatQuery(
          matcher.name,
          operator,
          matcher.isRegex ? `^${value}$` : value
        )
      );

      if (silenceFormStore.data.alertmanagers.length > 1) {
        filters.push(
          FormatQuery(
            StaticLabels.AlertManager,
            QueryOperators.Regex,
            `^(${silenceFormStore.data.alertmanagers
              .map(am => am.label)
              .join("|")})$`
          )
        );
      } else if (silenceFormStore.data.alertmanagers.length === 1) {
        filters.push(
          FormatQuery(
            StaticLabels.AlertManager,
            QueryOperators.Equal,
            silenceFormStore.data.alertmanagers[0].label
          )
        );
      }

      const alertsURI =
        FormatBackendURI("alerts.json?") + FormatAlertsQ(filters);

      this.matchedAlerts.fetch = fetch(alertsURI, { credentials: "include" })
        .then(result => {
          return result.json();
        })
        .then(result => {
          this.matchedAlerts.setTotal(result.totalAlerts);
          this.matchedAlerts.setError(null);
        })
        .catch(err => {
          console.trace(err);
          return this.matchedAlerts.setError(err.message);
        });
    }, 300);

    onUpdateCounter = () => {
      const { matcher } = this.props;

      if (matcher.name === "" || matcher.values.length === 0) {
        this.matchedAlerts.setTotal(0);
        this.matchedAlerts.setError(null);
        return;
      }

      this.onFetch();
    };

    componentDidMount() {
      this.onUpdateCounter();
    }

    componentDidUpdate() {
      this.onUpdateCounter();
    }

    render() {
      const { silenceFormStore, matcher } = this.props;

      if (this.matchedAlerts.error !== null) {
        return (
          <FontAwesomeIcon className="text-danger" icon={faExclamationCircle} />
        );
      }

      const matcherHash = hash({
        alertmanagers: silenceFormStore.data.alertmanagers,
        matcher: {
          name: matcher.name,
          values: matcher.values,
          isRegex: matcher.isRegex
        }
      });

      return (
        <span
          className="badge badge-light badge-pill"
          style={{ fontSize: "85%" }}
          data-hash={matcherHash}
        >
          {this.matchedAlerts.total}
        </span>
      );
    }
  }
);

export { MatchCounter };
