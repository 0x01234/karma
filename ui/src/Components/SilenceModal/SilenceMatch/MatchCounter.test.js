import React from "react";

import { mount } from "enzyme";

import toDiffableHtml from "diffable-html";

import { NewEmptyMatcher, MatcherValueToObject } from "Stores/SilenceFormStore";
import { MatchCounter } from "./MatchCounter";

let matcher;

beforeEach(() => {
  fetch.resetMocks();

  matcher = NewEmptyMatcher();
});

afterEach(() => {
  jest.restoreAllMocks();
});

const MountedMatchCounter = () => {
  return mount(<MatchCounter matcher={matcher} />);
};

describe("<MatchCounter />", () => {
  it("matches snapshot with empty matcher", () => {
    const tree = MountedMatchCounter();
    expect(toDiffableHtml(tree.html())).toMatchSnapshot();
  });

  it("logs a trace on failed fetch", async () => {
    const consoleSpy = jest
      .spyOn(console, "trace")
      .mockImplementation(() => {});
    fetch.mockReject("Fetch error");

    // we need to set name & value to trigger fetch
    matcher.name = "foo";
    matcher.values = [MatcherValueToObject("bar")];

    const tree = MountedMatchCounter();
    await expect(tree.instance().matchedAlerts.fetch).resolves.toBeUndefined();
    expect(consoleSpy).toHaveBeenCalled();
  });

  it("renders error icon on failed fetch", async () => {
    jest.spyOn(console, "trace").mockImplementation(() => {});
    fetch.mockReject("Fetch error");

    // we need to set name & value to trigger fetch
    matcher.name = "foo";
    matcher.values = [MatcherValueToObject("bar")];

    const tree = MountedMatchCounter();
    await expect(tree.instance().matchedAlerts.fetch).resolves.toBeUndefined();
    expect(toDiffableHtml(tree.html())).toMatch(/exclamation-circle/);
  });

  it("totalAlerts is 0 after mount", async () => {
    const tree = MountedMatchCounter();
    expect(tree.text()).toBe("0");
  });

  it("updates totalAlerts after successful fetch", async () => {
    fetch.mockResponse(JSON.stringify({ totalAlerts: 123 }));

    // we need to set name & value to trigger fetch
    matcher.name = "foo";
    matcher.values = [MatcherValueToObject("bar")];

    const tree = MountedMatchCounter();
    expect(tree.text()).toBe("0");
    await expect(tree.instance().matchedAlerts.fetch).resolves.toBeUndefined();
    expect(tree.text()).toBe("123");
  });

  it("sends correct query string for a 'foo=bar' matcher", async () => {
    fetch.mockResponse(JSON.stringify({ totalAlerts: 0 }));

    matcher.name = "foo";
    matcher.values = [MatcherValueToObject("bar")];
    matcher.isRegex = false;

    const tree = MountedMatchCounter();
    await expect(tree.instance().matchedAlerts.fetch).resolves.toBeUndefined();
    expect(fetch.mock.calls[0][0]).toBe("./alerts.json?q=foo%3Dbar");
  });

  it("sends correct query string for a 'foo=~bar' matcher", async () => {
    fetch.mockResponse(JSON.stringify({ totalAlerts: 0 }));

    matcher.name = "foo";
    matcher.values = [MatcherValueToObject("bar")];
    matcher.isRegex = true;

    const tree = MountedMatchCounter();
    await expect(tree.instance().matchedAlerts.fetch).resolves.toBeUndefined();
    expect(fetch.mock.calls[0][0]).toBe("./alerts.json?q=foo%3D~%5Ebar%24");
  });

  it("sends correct query string for a 'foo=~(bar|baz)' matcher", async () => {
    fetch.mockResponse(JSON.stringify({ totalAlerts: 0 }));

    matcher.name = "foo";
    matcher.values = [MatcherValueToObject("bar"), MatcherValueToObject("baz")];
    matcher.isRegex = true;

    const tree = MountedMatchCounter();
    await expect(tree.instance().matchedAlerts.fetch).resolves.toBeUndefined();
    expect(fetch.mock.calls[0][0]).toBe(
      "./alerts.json?q=foo%3D~%5E%28bar%7Cbaz%29%24"
    );
  });
});
