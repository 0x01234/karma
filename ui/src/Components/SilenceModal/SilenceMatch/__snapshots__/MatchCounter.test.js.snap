// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<MatchCounter /> matches snapshot with empty matcher 1`] = `
"
<div class
     style=\\"display: inline-block; max-width: 100%;\\"
     data-tooltipped
     aria-describedby=\\"tippy-tooltip-1\\"
     data-original-title=\\"Number of alerts matching this label\\"
>
  <span class=\\"badge badge-light badge-pill\\"
        style=\\"font-size: 85%;\\"
  >
    0
  </span>
</div>
"
`;
